libp2p/NodeTable.cpp

```diff
@@ -69,79 +69,81 @@ void NodeTable::processEvents()
 
 bool NodeTable::addNode(Node const& _node)
 {
-    shared_ptr<NodeEntry> entry;
+    LOG(m_logger) << "Adding node " << _node;
+
+    if (!isValidNode(_node))
+        return false;
+
+    bool needToPing = false;
     DEV_GUARDED(x_nodes)
     {
         auto const it = m_allNodes.find(_node.id);
-        if (it == m_allNodes.end())
-            entry = createNodeEntry(_node, 0, 0);
-        else
-            entry = it->second;
+        needToPing = (it == m_allNodes.end() || !it->second->hasValidEndpointProof());
```

Comment:
Should we also check m_sentPings? What if we sent a ping recently but just haven't received a response yet?

Reply:
yeah good point, it would improve things, but unfortunately it's not easy to do it here, because `m_sentPings` is accessed only from the network thread...

Reply:
We could either make `addNode` asynchronous; or we could create another private method for adding nodes from the packet handlers (but it would be very similar to `addNode`); or maybe we could check before sending Ping in `ping()` whether the node already happened to be validated, then skip additional `Ping`.

All this seems quite ugly for me for now, I'd suggest first to observe how often we get double pings because of this problem, then decide whether it make sense to complicate the code...